### PR TITLE
Support STORE_ATTR and STORE_SUBSCR

### PR DESCRIPTION
**Pull request by alexmojaki**
@pwwang

### CODE REVIEW

**Comment by alexmojaki** [comment]
You will have to figure out why the test is failing and how to fix it

**Comment by alexmojaki** [comment]
I'm suddenly realising this behaviour is a bit confusing. Code like `self.foo, self.bar = self.pair = spam()` will always identify the full `ast.Assign` statement, instead of one of the three `ast.Attribute`s. How do I use this to know what's being assigned when I'm inside `__setattr__`?

**Comment by pwwang** [comment]
> I'm suddenly realising this behaviour is a bit confusing. Code like `self.foo, self.bar = self.pair = spam()` will always identify the full `ast.Assign` statement, instead of one of the three `ast.Attribute`s. How do I use this to know what's being assigned when I'm inside `__setattr__`?

True. It's tough to distinguish in the above situation. I actually have it in mind at the beginning. However, it doesn't disobey this package's aim, which is to find the correct "node". In this case, `self.foo, self.bar = self.pair = spam()` is just a single assignment node with multiple targets. Is it the correct node? Yes. As for how to deal with the node, I think we could leave it to the downstream developer. One could either restrict the number of targets to only allow single "=" in the node, or distinguish it from the attribute name (`foo` vs `bar`) and raise an error only when there are attributes with the same name.

**Comment by alexmojaki** [comment]
> Is it the correct node? Yes.

But it's just giving the statement. That was already pretty easy to do with the package. From within `spam()`, the correct node is the `ast.Call`. Giving the containing statement would be silly. From within `__setattr__`, only one of `self.foo` etc is actually being executed. It doesn't assign all three at once.

3.11 is able to identify the exact node:

```
    x[0], x[1] = 1, 2
          ~^^^
IndexError: list assignment index out of range
```

So `executing` should be able to do so too in 3.11 once #31 is merged. And this kind of behaviour should be consistent accross Python versions.

I do think it's possible to identify `self.foo` with roughly the same trick as usual by replacing it with `(self ** sentinel).foo`. So this might not be that tricky. If you don't want to do that, then I think `node` should be either `self.foo` or `None` when there are multiple targets (including unpacking) so that the type of node identified doesn't change across versions.

**Comment by pwwang** [comment]
> From within `spam()`, the correct node is the `ast.Call`

But what we expected is a `__setattr__()` or `__setitem__()` call, which is `ast.Assign`. For `self.foo = self.bar = 1`, inside `__setattr__()`, we are expecting an Assign node, but not only `self.foo` or `self.bar`, which is a part of the Assign node. Even if we have `self.foo, self.bar = 1, 2`, The Assign node is still what we expected. It's just a little more complicated node:

```
Assign(
  targets=[
    Tuple(
      elts=[
        Attribute(value=Name(id='self', ctx=Load()), attr='foo', ctx=Store()), 
        Attribute(value=Name(id='self', ctx=Load()), attr='bar', ctx=Store())
      ], 
      ctx=Store()
    )
  ], 
  value=Tuple(elts=[Constant(value=1), Constant(value=2)], ctx=Load())
)
```


A runnable example with this PR:

```python
import ast
import sys
import executing

class A:
    def __setattr__(self, name, value):
        frame = sys._getframe(1)
        node = executing.Source.executing(frame).node
        print(ast.dump(node))
        # Deal with the node further

a = A()
a.foo, (b, a.bar), a.baz = 1, (2, 3), 4
```
prints:

```
Assign(targets=[Tuple(elts=[Attribute(value=Name(id='a', ctx=Load()), attr='foo', ctx=Store()), Tuple(elts=[Name(id='b', ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='bar', ctx=Store())], ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='baz', ctx=Store())], ctx=Store())], value=Tuple(elts=[Constant(value=1), Tuple(elts=[Constant(value=2), Constant(value=3)], ctx=Load()), Constant(value=4)], ctx=Load()))
Assign(targets=[Tuple(elts=[Attribute(value=Name(id='a', ctx=Load()), attr='foo', ctx=Store()), Tuple(elts=[Name(id='b', ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='bar', ctx=Store())], ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='baz', ctx=Store())], ctx=Store())], value=Tuple(elts=[Constant(value=1), Tuple(elts=[Constant(value=2), Constant(value=3)], ctx=Load()), Constant(value=4)], ctx=Load()))
Assign(targets=[Tuple(elts=[Attribute(value=Name(id='a', ctx=Load()), attr='foo', ctx=Store()), Tuple(elts=[Name(id='b', ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='bar', ctx=Store())], ctx=Store()), Attribute(value=Name(id='a', ctx=Load()), attr='baz', ctx=Store())], ctx=Store())], value=Tuple(elts=[Constant(value=1), Tuple(elts=[Constant(value=2), Constant(value=3)], ctx=Load()), Constant(value=4)], ctx=Load()))
```

**Comment by alexmojaki** [comment]
No, I think I would be expecting one of the `Attribute` nodes. If I wanted the `Assign` I could just use `Source.executing(frame).statements` and that would almost always work.

If I run your same example code with #31 and 3.11, I get:

```python
Attribute(value=Name(id='a', ctx=Load()), attr='foo', ctx=Store())
Attribute(value=Name(id='a', ctx=Load()), attr='bar', ctx=Store())
Attribute(value=Name(id='a', ctx=Load()), attr='baz', ctx=Store())
```

**Comment by pwwang** [comment]
That looks like results from `__getattr__()` instead of `__setattr__()`.

**Comment by alexmojaki** [comment]
The difference is the `ctx=Store()`.

It also works when assigning in a loop or comprehension, like `for a.foo in ...`.

`executing` is used to highlight nodes in tracebacks, notably in IPython. It should highlight the same thing that's already highlighted by 3.11 by default, if it can highlight anything at all.

I'm firm on this. `node` should be `None` or an `Attribute` or `Subscript`.

**Comment by pwwang** [comment]
Alright. Closing.

**Comment by alexmojaki** [comment]
As I said, it's acceptable to set `node` to the only `Attribute` or `Subscript` on the left of the `=` if there's only one, and leave it as `None` otherwise. Partial support without always being to identify the exact node is fine.

File: tests/test_main.py
Comment (alexmojaki) [inline tests/test_main.py:604]: I don't think this is needed. Checking the attribute name should generally allow disambiguation. But using `in` isn't specific enough, and `==` fails when private attributes are used.

Comment (pwwang) [reply tests/test_main.py:604]: But it failed for `xml.dom.minidom` with statement: `self._attrs, self._attrsNS, self._ownerElement = state`

Comment (alexmojaki) [reply tests/test_main.py:604]: Yes, I'm saying a smarter check on attribute names would be able to handle that.

Comment (pwwang) [reply tests/test_main.py:604]: Not sure I followed you. Mind working on it?

**Review by pwwang — APPROVED** [review]
Looks good to me.

**Comment by alexmojaki** [comment]
Released 0.9.0

**Pull request merged**
